### PR TITLE
Changes to allow ceph-ansible and vagrant to work on Openstack VMs

### PR DESCRIPTION
**Pull request by jimcurtis**
This has been tested with our internal OS1 Openstack VMs.  It allows you to use vagrant up --provider=openstack to create Openstack VMs and start ceph containers on them.  It has been tested with various OS1 images including centos, centos atomic and rhel atomic.

### CODE REVIEW

File: Vagrantfile.openstack [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# -*- mode: ruby -*-
```
Comment (leseb) [inline Vagrantfile.openstack]: Do you mind integrating this into the default Vagrantfile?
We currently support Virtualbox, VMware and Libvort providers within the same file so this shouldn't be an issue.

At some point, I'm hoping to have this branch merged with master so we should keep that in mind :)

File: roles/ceph-mds/tasks/docker/pre_requisite.yml [hunk 2]
```diff
@@ -19,6 +19,11 @@
     state: present
   when: ansible_os_family == 'RedHat'
 
+- name: install six
```
Comment (leseb) [inline roles/ceph-mds/tasks/docker/pre_requisite.yml]: can you use a `with_items` loop to install pip's packages? so you can incorporate that with the docker-py installation

File: roles/ceph-mon/tasks/docker/pre_requisite.yml [hunk 3]
```diff
@@ -47,6 +53,13 @@
   tags:
     with_pkg
 
+- name: install six
```
Comment (leseb) [inline roles/ceph-mon/tasks/docker/pre_requisite.yml]: can you use a `with_items` loop to install pip's packages? so you can incorporate that with the docker-py installation

File: roles/ceph-restapi/tasks/docker/pre_requisite.yml [hunk 4]
```diff
@@ -33,6 +33,13 @@
     state: started
     enabled: yes
 
+- name: install six
```
Comment (leseb) [inline roles/ceph-restapi/tasks/docker/pre_requisite.yml]: can you use a `with_items` loop to install pip's packages? so you can incorporate that with the docker-py installation

File: roles/ceph-rgw/tasks/docker/pre_requisite.yml [hunk 5]
```diff
@@ -29,6 +29,13 @@
   tags:
     with_pkg
 
+- name: install six
```
Comment (leseb) [inline roles/ceph-rgw/tasks/docker/pre_requisite.yml]: can you use a `with_items` loop to install pip's packages? so you can incorporate that with the docker-py installation

File: roles/ceph-mon/tasks/docker/start_docker_monitor.yml [hunk 6]
```diff
@@ -15,6 +15,7 @@
     image: "{{ ceph_mon_docker_username }}/{{ ceph_mon_docker_imagename }}"
     name: "{{ ansible_hostname }}"
     net: "host"
+    pid: "host"
```
Comment (leseb) [inline roles/ceph-mon/tasks/docker/start_docker_monitor.yml]: Just curiosity, why do you need to share the PID space with the host?

File: roles/ceph-mon/tasks/docker/pre_requisite.yml [hunk 7]
```diff
@@ -39,6 +39,12 @@
   tags:
     with_pkg
 
+- name: Pause after docker install before starting (on openstack vms)
```
Comment (leseb) [inline roles/ceph-mon/tasks/docker/pre_requisite.yml]: s/Pause/pause/

File: roles/ceph-osd/tasks/docker/start_docker_osd.yml [hunk 8]
```diff
@@ -1,4 +1,10 @@
 ---
+# For openstack VMs modify the mount point below depending on if the Openstack
+# VM deploy tool defaults to mounting ephemeral disks
+- name: umount ceph disk (if on openstack)
+  shell: "umount /mnt"
```
Comment (leseb) [inline roles/ceph-osd/tasks/docker/start_docker_osd.yml]: can't you use the mount module (http://docs.ansible.com/ansible/mount_module.html) instead?

File: Vagrantfile.openstack (hunk 1 above)
Comment (jimcurtis) [inline Vagrantfile.openstack]: Yes.  I will go ahead and do that and update the PR.

File: roles/ceph-mds/tasks/docker/pre_requisite.yml (hunk 2 above)
Comment (jimcurtis) [inline roles/ceph-mds/tasks/docker/pre_requisite.yml]: Unfortunately, I need to ensure that a certain version of six is installed, so I can't use the with_items loop.  I did add a comment to indicate this.

File: roles/ceph-mon/tasks/docker/pre_requisite.yml (hunk 3 above)
Comment (jimcurtis) [inline roles/ceph-mon/tasks/docker/pre_requisite.yml]: Unfortunately, I need to ensure that a certain version of six is installed, so I can't use the with_items loop.  I did add a comment to indicate this.

File: roles/ceph-restapi/tasks/docker/pre_requisite.yml (hunk 4 above)
Comment (jimcurtis) [inline roles/ceph-restapi/tasks/docker/pre_requisite.yml]: Unfortunately, I need to ensure that a certain version of six is installed, so I can't use the with_items loop.  I did add a comment to indicate this.

File: roles/ceph-rgw/tasks/docker/pre_requisite.yml (hunk 5 above)
Comment (jimcurtis) [inline roles/ceph-rgw/tasks/docker/pre_requisite.yml]: Unfortunately, I need to ensure that a certain version of six is installed, so I can't use the with_items loop.  I did add a comment to indicate this.

File: roles/ceph-mon/tasks/docker/start_docker_monitor.yml (hunk 6 above)
Comment (jimcurtis) [inline roles/ceph-mon/tasks/docker/start_docker_monitor.yml]: It doesn't look like this is needed after all.  I will remove it.

File: roles/ceph-mon/tasks/docker/pre_requisite.yml (hunk 7 above)
Comment (jimcurtis) [inline roles/ceph-mon/tasks/docker/pre_requisite.yml]: Fixed.

File: roles/ceph-mds/tasks/docker/pre_requisite.yml (hunk 2 above)
Comment (leseb) [inline roles/ceph-mds/tasks/docker/pre_requisite.yml]: You can re-use this: https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-common/tasks/misc/system_tuning.yml#L19-L26

And get the vars like this:
https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-common/defaults/main.yml#L293-L298

So you should be able to have a single task with this.

File: roles/ceph-osd/tasks/docker/start_docker_osd.yml (hunk 8 above)
Comment (jimcurtis) [inline roles/ceph-osd/tasks/docker/start_docker_osd.yml]: Changed to use mount module

File: vagrant_variables.yml.openstack [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+disks: "[ '/dev/vdb' ]"
+
+# For Openstack VMs, the lan is usually eth0
+eth: 'eth0'
+
+# For Openstack VMs, choose the following box instead
+vagrant_box: 'openstack'
+
+# For Atomic (RHEL or Cento) uncomment the line below
+skip_tags: 'with_pkg'
```
Comment (rootfs) [inline vagrant_variables.yml.openstack]: this is not consistent with the comment :)

**Comment by rootfs** [comment]
please rebase

**Comment by rootfs** [comment]
Please add instruction to `README.md` on how to install openstack plugin and run vagrant with openstack plugin.

File: Vagrantfile [hunk 10]
```diff
@@ -67,11 +78,30 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
...
   end
 
+  # OpenStack VMs
+  config.vm.provider :openstack do |os|
+    config.vm.synced_folder ".", "/home/#{USER}/vagrant", disabled: true
+    config.ssh.username = USER
+    config.ssh.private_key_path = settings['os_ssh_private_key_path']
+    config.ssh.pty = true
+    os.openstack_auth_url = settings['os_openstack_auth_url']
+    os.username = settings['os_username']
```
Comment (rootfs) [inline Vagrantfile]: make it also able to accept ENV, like
`os.username     = ENV['OS_USERNAME']
os.password     = ENV['OS_PASSWORD']`

**Comment by jimcurtis** [comment]
Completed latest rebase.

File: Vagrantfile [hunk 11]
```diff
@@ -2,6 +2,7 @@
 # vi: set ft=ruby :
 
 require 'yaml'
+require 'vagrant-openstack-provider'
```
Comment (leseb) [inline Vagrantfile]: I assume vagrant will fail if not present, can we have a condition for this?

File: roles/ceph-mon/tasks/docker/pre_requisite.yml [hunk 12]
```diff
@@ -39,6 +39,22 @@
   tags:
     with_pkg
 
+# NOTE (jimcurtis): need at least version 1.9.0 of six or we get:
+# re:NameError: global name 'DEFAULT_DOCKER_API_VERSION' is not defined
+
```
Comment (leseb) [inline roles/ceph-mon/tasks/docker/pre_requisite.yml]: space not needed

File: roles/ceph-osd/tasks/docker/pre_requisite.yml [hunk 13]
```diff
@@ -37,6 +37,22 @@
   tags:
     with_pkg
 
+# NOTE (jimcurtis): need at least version 1.9.0 of six or we get:
+# re:NameError: global name 'DEFAULT_DOCKER_API_VERSION' is not defined
+
```
Comment (leseb) [inline roles/ceph-osd/tasks/docker/pre_requisite.yml]: new line not needed

File: roles/ceph-osd/tasks/docker/start_docker_osd.yml [hunk 14]
```diff
@@ -1,4 +1,10 @@
 ---
+# For openstack VMs modify the mount point below depending on if the Openstack
+# VM deploy tool defaults to mounting ephemeral disks
+- name: umount ceph disk (if on openstack)
+  mount: name=/mnt src=/dev/vdb fstype=ext3 state=unmounted
```
Comment (leseb) [inline roles/ceph-osd/tasks/docker/start_docker_osd.yml]: please use the following yaml syntax:

```
mount:
  name:
  src:
...
```

File: roles/ceph-restapi/tasks/docker/pre_requisite.yml [hunk 15]
```diff
@@ -27,6 +27,16 @@
     state: present
   when: ansible_os_family == 'RedHat'
 
+# NOTE (jimcurtis): need at least version 1.9.0 of six or we get:
+# re:NameError: global name 'DEFAULT_DOCKER_API_VERSION' is not defined
+
```
Comment (leseb) [inline roles/ceph-restapi/tasks/docker/pre_requisite.yml]: new line not needed

File: roles/ceph-rgw/tasks/docker/pre_requisite.yml [hunk 16]
```diff
@@ -35,6 +35,16 @@
     state: present
   when: ansible_os_family == 'RedHat'
 
+# NOTE (jimcurtis): need at least version 1.9.0 of six or we get:
+# re:NameError: global name 'DEFAULT_DOCKER_API_VERSION' is not defined
+
```
Comment (leseb) [inline roles/ceph-rgw/tasks/docker/pre_requisite.yml]: new line not needed

File: Vagrantfile [hunk 17]
```diff
@@ -92,8 +122,9 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   (0..NRGWS - 1).each do |i|
     config.vm.define "rgw#{i}" do |rgw|
       rgw.vm.hostname = "ceph-rgw#{i}"
-      rgw.vm.network :private_network, ip: "#{SUBNET}.4#{i}"
-
+      if !OSVM
+        rgw.vm.network :private_network, ip: "#{SUBNET}.4#{i}"
```
Comment (leseb) [inline Vagrantfile]: can you rebase your patch again? so you get this fix: https://github.com/ceph/ceph-ansible/pull/535

File: vagrant_variables.yml.openstack [hunk 18]
```diff
@@ -0,0 +1,37 @@
+---
```
Comment (leseb) [inline vagrant_variables.yml.openstack]: can the content of integrated into the default `vagrant_variables.yml.sample` instead?

File: vagrant_variables.yml.openstack (hunk 9 above)
Comment (jimcurtis) [inline vagrant_variables.yml.openstack]: Made change to make it consistent.

File: Vagrantfile (hunk 10 above)
Comment (jimcurtis) [inline Vagrantfile]: This is more appropriate in the vagrant_variables.yml.openstack file.  I made the change there.

File: Vagrantfile (hunk 11 above)
Comment (jimcurtis) [inline Vagrantfile]: Put the require statement inside the condition as requested.

File: roles/ceph-mon/tasks/docker/pre_requisite.yml (hunk 12 above)
Comment (jimcurtis) [inline roles/ceph-mon/tasks/docker/pre_requisite.yml]: removed extra line

File: roles/ceph-osd/tasks/docker/pre_requisite.yml (hunk 13 above)
Comment (jimcurtis) [inline roles/ceph-osd/tasks/docker/pre_requisite.yml]: removed extra line

File: roles/ceph-osd/tasks/docker/start_docker_osd.yml (hunk 14 above)
Comment (jimcurtis) [inline roles/ceph-osd/tasks/docker/start_docker_osd.yml]: Changed to follow recommended syntax

File: roles/ceph-restapi/tasks/docker/pre_requisite.yml (hunk 15 above)
Comment (jimcurtis) [inline roles/ceph-restapi/tasks/docker/pre_requisite.yml]: removed line

File: roles/ceph-rgw/tasks/docker/pre_requisite.yml (hunk 16 above)
Comment (jimcurtis) [inline roles/ceph-rgw/tasks/docker/pre_requisite.yml]: removed line

File: Vagrantfile (hunk 17 above)
Comment (jimcurtis) [inline Vagrantfile]: added that fix

File: vagrant_variables.yml.openstack (hunk 18 above)
Comment (jimcurtis) [inline vagrant_variables.yml.openstack]: It is possible.  But I figured that it was different enough that not having a bunch of conditional stuff would be cleaner.  We can always look into this in the future if things get unwieldy.

**Comment by leseb** [comment]
@jimcurtis can you please push your new changes?

**Comment by jimcurtis** [comment]
Pushed the changes.

----- Original Message -----

File: README.md [hunk 19]
```diff
@@ -151,7 +151,39 @@ VBoxManage storagectl `VBoxManage list vms |grep ceph-ansible_osd0|awk '{print $
 
 then run `vagrant up --provider=virtualbox` again.
 
-
+## Setup for Vagrant using OpenStack provider
+
+Install the Vagrant plugin for the openstack provider: vagrant plugin install vagrant-openstack-provider.
+
+cp site.yml.sample site.yml
```
Comment (leseb) [inline README.md]: Make this a code block please using backquotes.

Comment (jimcurtis) [inline README.md]: Made recommended changes.

----- Original Message -----

File: README.md [hunk 20]
```diff
@@ -151,7 +151,44 @@ VBoxManage storagectl `VBoxManage list vms |grep ceph-ansible_osd0|awk '{print $
...
+```bash
+$ cp site.yml.sample site.yml
+$ cp group_vars/all.docker.sample group_vars/all
+$ cp vagrant_variables.yml.openstack vagrant_variables.yml
+```
+* Edit `vagrant_variables.yml`:
+  Set `mon_vms` and `osd_vms` to the numbers you want.
+  If you are using an Atomic image, un-comment out the `skip_tags` line.
+  Un-comment the `os_` lines.
+    Set `os_ssh_username` to 'centos' for Centos and 'cloud-user' for
```
Comment (leseb) [inline README.md]: Can you change the indentation here?

**Pull request merged**